modules/react-native-status/android/src/main/java/im/status/ethereum/module/StatusModule.java

```diff
@@ -117,11 +122,113 @@ public String getNoBackupDirectory() {
       return this.getReactApplicationContext().getNoBackupFilesDir().getAbsolutePath();
     }
 
-    public void handleSignal(String jsonEvent) {
-        Log.d(TAG, "Signal event: " + jsonEvent);
-        WritableMap params = Arguments.createMap();
-        params.putString("jsonEvent", jsonEvent);
-        this.getReactApplicationContext().getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class).emit("gethEvent", params);
+    private static WritableMap convertJsonToMap(JSONObject jsonObject) throws JSONException {
+        WritableMap map = new WritableNativeMap();
+        
+        Iterator<String> iterator = jsonObject.keys();
+        while (iterator.hasNext()) {
+            String key = iterator.next();
+            Object value = jsonObject.get(key);
+            if (value instanceof JSONObject) {
+                map.putMap(key, convertJsonToMap((JSONObject) value));
+            } else if (value instanceof  JSONArray) {
+                map.putArray(key, convertJsonToArray((JSONArray) value));
+            } else if (value instanceof  Boolean) {
+                map.putBoolean(key, (Boolean) value);
+            } else if (value instanceof  Integer) {
+                map.putInt(key, (Integer) value);
+            } else if (value instanceof  Double) {
+                map.putDouble(key, (Double) value);
+            } else if (value instanceof String)  {
+                map.putString(key, (String) value);
+            } else if (value == JSONObject.NULL){
+                map.putNull(key);
+            } else {
+                map.putString(key, value.toString());
+            }
+        }
+        return map;
+    }
+    
+    private static WritableArray convertJsonToArray(JSONArray jsonArray) throws JSONException {
+        WritableArray array = new WritableNativeArray();
+        
+        for (int i = 0; i < jsonArray.length(); i++) {
+            Object value = jsonArray.get(i);
+            if (value instanceof JSONObject) {
```

Comment:
Not like it's very important, but `instanceof` is kind of a code smell and you would probably define a separate method for each type. Also `instanceof` should be a bit slower than dispatching, though not sure it will have a serious impact here.

Reply:
@rasom thanks for the review, maybe I should get rid of this, it's just some code I got from a blog article, I wanted to test passing js objects directly vs parse json on react side. It saves a few ms but I don't know if it is worth it. I'm very rusty in Java how would you replace that?

Reply:
Probably by defining a separate `convert` method for each type and then allow it to choose the right one based on the type. 
```java
private static void convert(WritableArray arr, JSONObject obj) {
    array.pushMap(convertJsonToMap((JSONObject) obj));
}
...
private static void convert(WritableArray arr, Object obj) {
    array.pushString(obj.toString());
}

//and then in convertJsonToArray
...
for (int i = 0; i < jsonArray.length(); i++) {
     convert(array, jsonArray.get(i));
}
...
```
probably something like this will work